lib/web/terminal.go

```diff
@@ -284,12 +292,44 @@ func (t *TerminalHandler) makeClient(ws *websocket.Conn) (*client.TeleportClient
 	tc.OnShellCreated = func(s *ssh.Session, c *ssh.Client, _ io.ReadWriteCloser) (bool, error) {
 		t.sshSession = s
 		t.windowChange(&t.params.Term)
+
 		return false, nil
 	}
 
 	return tc, nil
 }
 
+// startPingLoop starts a loop that will continuously send a ping frame through the websocket
+// to prevent the connection between web client and teleport proxy from becoming idle.
+// Interval is determined by the keep_alive_interval config set by user (or default).
+// Loop will terminate when there is an error sending ping frame or when terminal session is closed.
+func (t *TerminalHandler) startPingLoop(ws *websocket.Conn) {
+	// Define our own marshal func to just return a ping payload type.
+	codec := websocket.Codec{Marshal: func(v interface{}) (data []byte, payloadType byte, err error) {
+		return nil, websocket.PingFrame, nil
+	}}
+
+	t.log.Debugf("Starting websocket ping loop with interval %v.", t.params.KeepAliveInterval)
+	tickerCh := time.NewTicker(t.params.KeepAliveInterval)
+	defer tickerCh.Stop()
+
+	for {
+		select {
+		case <-tickerCh.C:
+			// Pongs are internally handled by the websocket library.
+			// https://github.com/golang/net/blob/master/websocket/hybi.go#L291
+			if err := codec.Send(ws, nil); err != nil {
+				t.log.Errorf("Unable to send ping frame to web client: %v.", err)
+				t.Close()
```

Comment:
Why should we terminate the connection if for some reason ping failed? Ping/pongs frames do not resurface on the application layer so it really does not matter. Why cannot we log an error and exit this go-routine instead? That would make a connection more more resilient to errors.

Reply:
Philosophical difference I think.  I'm generally of the opinion that components should always fail-fast unless there is a specific reason not to, and definitely shouldn't enter states of permanent partial failure.  A component which "does thing X" is much easier to reason about than a component that "might do thing X unless something went wrong at some point in the past".  The difference is inconsequential for most individual cases but, at scale, it makes a huge difference in terms of how difficult it is to reason about the wider system.

As far as I'm aware, a `ping` operation *should* always succeed.  If it didn't, something is either wrong with the connection, or wrong with our code.  Neither of these are spurious failures.

If there is a concern of spurious errors of some kind, then the ping loop should re-attempt on such errors, not exit.

Reply:
Hm...to me this is not philosophical but rather practical. We are adding this extra feature so load-balancers (like cloudflare) do not terminate a connection. This business is between us and loadbalancers and I do not see a reason why a user should be impacted if we fail to send `ping` frame.  To me it would be good enough if we exit the loop and let it be so a user who is switching a network in between wont be impacted. 

Again, failed ping event is not critical to kick a user from the terminal. If failed, please exit the loop and stop even trying but keep my terminal open as it has "0" impact on me. Should we keep re-trying for some time? I think yes but what we have now is good enough. But please, do not terminate it.

As for philosophical differences, "fail-fast" !== fragile connections.

Reply:
The reason for this PR is to add pings which will prevent websocket connections from being dropped by load balancers. If we fail to send one ping and stop trying, then we're immediately back in the previous situation where the connection will subsequently be dropped after ~100 seconds of inactivity. 

I agree that we shouldn't close the connection if one ping fails - we need ping frames to be sent in a constant loop to prevent this behaviour. We should retry infinitely on the same schedule. Eventually, if the pings keep failing, then the connection will be closed anyway (probably due to poor connectivity) which is something outside of our control.

Reply:
Let's consider _why_ a ping might fail.
Since this runs over TCP, it wouldn't be cause by packet loss; TCP masks that. And any larger transient network error should break the connection.
It can have something to do with encoding a frame (bug in websocket library, memory corruption, etc).

So a failed ping indicates one of:
- broken connection - in this case it's in teleport's interest to free up server-side resources as soon as possible
- memory or encoding problem - this is tricky; on the one hand, we can stick to Postel's law and do our best to keep the session going; on the other hand, there's something seriously wrong and failing fast and loud will attract human attention

> We should retry infinitely on the same schedule. Eventually, if the pings keep failing, then the connection will be closed anyway (probably due to poor connectivity) which is something outside of our control.

But we should avoid trying to send pings if the connection was closed. Otherwise, it's a short-term resource leak.

Reply:
> But we should avoid trying to send pings if the connection was closed. Otherwise, it's a short-term resource leak.

@awly this is why I think exiting the loop would be good enough. This would terminate this feature and as Gus pointed out the rest would fallback to default behavior.

Reply:
per @russjones, I removed terminating connection from error sending pings

Reply:
Some update on the issue, we've chatted with Andrew and it's OK to _close_ a connection in that loop. There is no harm in just existing the loop but there is no point of keeping the dead connection around (or re-trying) after this point. Thank you guys for making it more clear!

Reply:
reverted back to terminating connection with failed pings